server/src/main/java/org/elasticsearch/snapshots/RepositoryIntegrityHealthIndicatorService.java

```diff
@@ -73,78 +104,160 @@ public String name() {
-                List.of(ImpactArea.BACKUP)
-            )
-        );
+        var repositoryHealthAnalyzer = new RepositoryHealthAnalyzer(clusterState, repositories, healthInfo.repositoriesInfoByNode());
         return createIndicator(
-            RED,
-            createCorruptedRepositorySummary(corrupted),
-            verbose
-                ? new SimpleHealthIndicatorDetails(
-                    Map.of(
-                        "total_repositories",
-                        totalRepositories,
-                        "corrupted_repositories",
-                        corruptedRepositories,
-                        "corrupted",
-                        limitSize(corrupted, 10)
-                    )
-                )
-                : HealthIndicatorDetails.EMPTY,
-            impacts,
-            List.of(
-                new Diagnosis(
-                    CORRUPTED_REPOSITORY,
-                    List.of(new Diagnosis.Resource(Type.SNAPSHOT_REPOSITORY, limitSize(corrupted, maxAffectedResourcesCount)))
-                )
-            )
+            repositoryHealthAnalyzer.getHealthStatus(),
+            repositoryHealthAnalyzer.getSymptom(),
+            repositoryHealthAnalyzer.getDetails(verbose),
+            repositoryHealthAnalyzer.getImpacts(),
+            repositoryHealthAnalyzer.getDiagnoses(maxAffectedResourcesCount)
         );
     }
 
-    private static String createCorruptedRepositorySummary(List<String> corrupted) {
-        var message = new StringBuilder().append("Detected [").append(corrupted.size()).append("] corrupted snapshot repositories: ");
-        collectionToDelimitedStringWithLimit(corrupted, ",", "[", "].", 1024, message);
-        return message.toString();
+    /**
+     * Analyzer for the cluster's repositories health; aids in constructing a {@link HealthIndicatorResult}.
+     */
+    static class RepositoryHealthAnalyzer {
+        private final ClusterState clusterState;
+        private final int totalRepositories;
+        private final List<String> corruptedRepositories;
+        private final Set<String> unknownRepositories = new HashSet<>();
+        private final Set<String> nodesWithUnknownRepos = new HashSet<>();
+        private final Set<String> invalidRepositories = new HashSet<>();
+        private final Set<String> nodesWithInvalidRepos = new HashSet<>();
+        private final HealthStatus healthStatus;
+
+        private RepositoryHealthAnalyzer(
+            ClusterState clusterState,
+            List<RepositoryMetadata> repositories,
+            Map<String, RepositoriesHealthInfo> repositoriesHealthByNode
+        ) {
+            this.clusterState = clusterState;
+            this.totalRepositories = repositories.size();
+            this.corruptedRepositories = repositories.stream()
+                .filter(repository -> repository.generation() == RepositoryData.CORRUPTED_REPO_GEN)
+                .map(RepositoryMetadata::name)
+                .toList();
```

Comment:
We use this list in two places (potentially) with `limitSize()`. If we sort it here then it can be nicely stable in both places.